src/rdiff-backup-delete

```diff
@@ -225,23 +226,45 @@ def _remove_increments(path, dry_run):
 
 def _rmtree(path, dry_run):
     """
-    Custom implementation of shutil.rmtree() to support deletion of symlink.
+    Custom implementation of shutil.rmtree() to handle permission errors,
+    fifo and symlink deletion.
     """
+    if dry_run:
+        return
+
+    # Try to change the permissions of the file or directory to delete them.
+    def onerror(func, path, exc_info):
+        if exc_info[0] == PermissionError:
+            # Parent directory must allow rwx
+            if not os.access(os.path.dirname(path), os.W_OK | os.R_OK | os.X_OK):
+                os.chmod(os.path.dirname(path), 0o0700)
+            if not os.access(path, os.W_OK | os.R_OK):
+                os.chmod(path, 0o0600)
+            return func(path)
+        raise
+
     try:
         mode = os.lstat(path).st_mode
-    except os.error:
+    except FileNotFoundError:
+        # Nothing to delete
         mode = 0
+    except OSError:
```

Comment:
why don't you directly `except PermissionError` and remove the check from the `on_error` function? Then you wouldn't even need to pass the exception to the function.

Reply:
Was previously handling a different kind or error in an intermediate version of the code. It become unnecessary. I agree.